tools: Include GMT offsets in timezone dropdown for better clarity (Fixes #20988).

Amaan-pathan opened the pull request:
## **settings: Display UTC offsets alongside timezones in dropdown.**

### **🔹 Overview**
This PR improves the timezone selection experience by showing GMT offsets next to location names.  
Users can now select a timezone based on its offset rather than just the location name.

### **🔹 Changes Made**
- Updated `tools/setup/build_timezone_values` to fetch and display GMT offsets.
- Modified `web/templates/settings/profile_settings.hbs` to include UTC offsets in the dropdown.
- Timezones are now sorted and stored in `timezones.json`, including their offsets.

### **🔹 Technical Details**
- Introduced `get_gmt_offset(tz_name)`, which calculates offsets using Python’s `zoneinfo` and `datetime` modules.
- The dropdown now displays timezones in the format:
  - `Asia/Kolkata (GMT+5:30)`
  - `Asia/Muscat (GMT+4)`

### **📌 Before & After**
**Before:**
![image](https://github.com/user-attachments/assets/f6f3d3ff-ce2c-42dc-ae6e-e162ef2fe1f1)


**After:**
![image](https://github.com/user-attachments/assets/c8003502-6458-432e-943b-25370264d839)

### **✅ Fixes**
Fixes: #20988

### **📋 Testing & Considerations**
- [x] Verified UI changes manually.
- [x] Ensured no breaking changes were introduced.
- [x] Ran `tools/lint` to check for code style issues.
- [x] Ran `tools/test-backend` to confirm no test failures.

alya commented:
Please clean up your commit history and post again to request a review. See [here](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html) for guidelines.

Amaan-pathan commented:
@alya Thank you for the feedback. I will clean up the commit history as per the guidelines and post the update for review.

Aditya8840 commented on tools/setup/build_timezone_values:
```diff
@@ -3,22 +3,47 @@ import json
+        minutes = remainder // 60
+        sign = "+" if offset_seconds >= 0 else "-"
+
+        return f"GMT {sign}{int(hours):02}:{int(minutes):02}"
```
This could simplified to `return f"GMT({hours:+03d}:{minutes:02d})"`.

Amaan-pathan replied:
Thank you for the suggestion! That’s a great simplification. I'll update the code accordingly.

Aditya8840 commented on web/templates/settings/profile_settings.hbs:
```diff
@@ -27,7 +27,7 @@
 
                                 {{#each timezones}}
-                                    <option value="{{ this }}">{{ this }}</option>
+                                    <option value="{{ this.name }}">{{ this.name }} ({{ this.gmtOffset}})</option>
```
Do not use camelCase `gmtOffset`.

Amaan-pathan replied:
Thanks for pointing that out! I’ll update it to gmt_offset to match the convention.

Amaan-pathan committed:
settings: Show UTC offsets in timezone dropdown.

This improves usability by helping users quickly recognize timezones
with their offsets.

Fixes #20988.

Amaan-pathan commented:
> Please clean up your commit history and post again to request a review. See [here](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html) for guidelines.

I've cleaned up the commit history and made the necessary changes as per the guidelines. Please review it at your convenience. Thanks!

alya commented:
Let's format this more succinctly, like:

> GMT+9
> GMT+9:30

Amaan-pathan commented:
> Let's format this more succinctly, like:
> 
> > (GMT+9)
> > (GMT+9:30)

@alya Thanks for the feedback! I’ve updated the GMT offset format to match the expected style—GMT+9 for whole hours and GMT+9:30 where needed. Let me know if any further tweaks are required.

alya commented:
The PR description needs to be updated to demonstrate your changes.

Amaan-pathan commented:
> The PR description needs to be updated to demonstrate your changes.
@alya I've updated the PR description to clearly demonstrate the changes. Please review and let me know if any further improvements are needed.

alya commented:
@kuv2707 Would you perhaps be up for reviewing this one? Not a release goal.

Amaan-pathan commented:
Hey @kuv2707, just a gentle reminder to review this PR whenever you get a chance. Let me know if you need any clarifications. Thanks!

kuv2707 commented on web/templates/settings/profile_settings.hbs:
```diff
@@ -27,7 +27,7 @@
 
                                 {{#each timezones}}
-                                    <option value="{{ this }}">{{ this }}</option>
+                                    <option value="{{ this.name }}">{{ this.name }} ({{ this.gmt_offset}})</option>
```
This should be properly internationalized. Refer to https://zulip.readthedocs.io/en/latest/translating/internationalization.html#handlebars-templates

Amaan-pathan replied:
Thanks for the feedback! I’ve updated the code to follow Zulip’s internationalization guidelines by using {{t}} and passing variables correctly. Let me know if any further improvements are needed!

kuv2707 reviewed (commented):
Left some comments. Looks good mostly.

You may squash both commits into one. Each commit should be individually coherent.

Amaan-pathan commented on tools/setup/build_timezone_values:
```diff
@@ -3,22 +3,46 @@ import json
 import zoneinfo
+from datetime import datetime, timedelta
 
+# Define paths
```
Got it! I've removed the unnecessary comment.

Amaan-pathan commented on tools/setup/build_timezone_values:
```diff
@@ -3,22 +3,46 @@ import json
+        offset: timedelta | None = now.utcoffset()
+
+        if offset is None:
+            return "GMT(+00:00)"  # Default to UTC if offset is missing
```
I've removed the inline comment since the code is self-explanatory.

Amaan-pathan commented on tools/setup/build_timezone_values:
```diff
@@ -3,22 +3,46 @@ import json
+            return "GMT(+00:00)"  # Default to UTC if offset is missing
+
+        offset_seconds = offset.total_seconds()
+        hours, remainder = divmod(offset_seconds, 3600)  # Preserve sign directly
```
Got it! I've removed the comment.

Amaan-pathan commented on tools/setup/build_timezone_values:
```diff
@@ -3,22 +3,46 @@ import json
+    zoneinfo.available_timezones() - {"Factory", "localtime"} - set(get_canonical_timezone_map())
+)
+
+# Create a dictionary with GMT offsets
```
Got it! I've removed all the comments

Amaan-pathan commented on tools/setup/build_timezone_values:
```diff
@@ -3,22 +3,46 @@ import json
+
+# Save to JSON
 with open(OUT_PATH, "w") as f:
-    json.dump(
-        {
-            "timezones": sorted(
-                zoneinfo.available_timezones()
-                - {"Factory", "localtime"}
-                - set(get_canonical_timezone_map())
-            )
-        },
-        f,
-    )
```
I initially added indent=4 for better readability in the output and included the print statement for debugging purposes. However, I understand that these changes were not discussed beforehand. I’ve now reverted the code to its original state as per the feedback. Let me know if any further adjustments are needed!

kuv2707 replied:
If you wish to make minor amendments like these, you can add these in separate commits, such that they can simply be removed if the maintainers wish so.

Amaan-pathan replied:
Got it! I'll keep such minor amendments in separate commits next time for easier review. For now, I've reverted the changes to match the original state as per the feedback.

Amaan-pathan commented:
> Left some comments. Looks good mostly.
> 
> You may squash both commits into one. Each commit should be individually coherent.

Hey @kuv2707 I've made the necessary changes and merged the comments. Also, I have squashed both commits into one to ensure coherence. Let me know if any further modifications are needed!

kuv2707 commented on tools/setup/build_timezone_values:
```diff
@@ -11,14 +12,35 @@ from zerver.lib.timezone import get_canonical_timezone_map
+
+    except Exception as e:
+        print(f"Error processing {tz_name}: {e}")
+        return "GMT(+00:00)"
```
It would be better to return a string that more clearly demonstrates an error case, rather than one which is a valid string (for United Kingdom in this case)

Let's make it `GMT(?)`.

kuv2707 reviewed (approved):
Left minor suggestions. LGTM otherwise.

Amaan-pathan commented:
> Left minor suggestions. LGTM otherwise.

@kuv2707 Thanks for the review! I've addressed all the suggestions. Let me know if anything else needs to be changed.

kuv2707 commented:
@alya This looks good to me

alya commented:
Thanks @Amaan-pathan @kuv2707 ! Marking for integration review (not a release priority, so might wait a bit).

Amaan-pathan commented:
@alya Thank you for reviewing my PR! I want to work on the UI part as I have expertise in this area, and most UI issues are high-priority tasks. To get those assigned to me, it is essential for this PR to be merged as soon as possible. I would really appreciate your cooperation in expediting the process. Please let me know if any further changes are needed. Looking forward to your support!

timabbott commented on tools/setup/build_timezone_values:
```diff
@@ -11,14 +12,35 @@ from zerver.lib.timezone import get_canonical_timezone_map
+)
+
+
+timezone_data = [{"name": tz, "gmt_offset": get_gmt_offset(tz)} for tz in timezones]
```
I'm not sure why you're calling this GMT when the code is using UTC. Let's use UTC throughout, both in the code and the labels.

Amaan-pathan replied:
Hey @timabbott,Thanks for the clarification! I have updated all instances of GMT to UTC across both the code and UI labels to maintain consistency.

Earlier, I had used GMT as it was mentioned in the proposed menu in [issue #20988](https://github.com/zulip/zulip/issues/20988#issue-1118271507), but I’ve now aligned everything with UTC as per your feedback. Let me know if there's anything else you'd like me to adjust!

Amaan-pathan commented:
Hey @timabbott , I’ve updated the PR as per your feedback by replacing all instances of GMT with UTC for consistency. Let me know if you get a chance to review it or if there's anything else that needs adjustment. Thanks!

timabbott commented:
Merged, after bumping `PROVISION_VERSION`. Thanks for all the work on this @Amaan-pathan! I hope you learned a bunch.

Amaan-pathan commented:
> Merged, after bumping `PROVISION_VERSION`. Thanks for all the work on this @Amaan-pathan! I hope you learned a bunch.

Thanks! Learned a lot from this. Looking forward to contributing more!

andersk commented:
You have failed to account for daylight saving time and other time zone changes—UTC offsets are not constant, and it makes no sense to cache them.

- #39677
